File: internal/api/handlers.go
package api

import (
	"encoding/json"
	"io"

	"kb/internal"
	"kb/internal/retr"
)

// SearchHandler handles search requests and returns JSON results.
// This is a simple implementation that can be extended for HTTP serving.
func SearchHandler(dataDir string, topN int) func(io.Writer, string) {
	return func(w io.Writer, query string) {
		results, err := retr.HybridSearch(dataDir, query, topN)
		if err != nil {
			// In a real HTTP handler, this would be an error response
			results = []retr.SearchResult{}
		}
		
		// Convert to DocEntry format for compatibility
		docEntries := make([]internal.DocEntry, 0, len(results))
		for _, result := range results {
			docEntries = append(docEntries, result.Meta)
		}
		
		enc := json.NewEncoder(w)
		enc.SetIndent("", "  ")
		enc.Encode(docEntries)
	}
}

File: ts-source/src/internal/api/handlers.ts
import { DocEntry } from '../types.js';
import { hybridSearch } from '../retr/search.js';

export function createSearchHandler(dataDir: string, topN: number) {
  return async (query: string): Promise<DocEntry[]> => {
    try {
      const results = await hybridSearch(dataDir, query, topN);
      return results.map(result => result.metadata);
    } catch (error) {
      return [];
    }
  };
}

export function createExpressSearchHandler(dataDir: string, topN: number) {
  return async (req: any, res: any) => {
    try {
      const { query } = req.body;
      
      if (!query || typeof query !== 'string') {
        return res.status(400).json({ error: 'Query parameter is required' });
      }

      const results = await hybridSearch(dataDir, query, topN);
      const docEntries = results.map(result => result.metadata);
      
      res.json(docEntries);
    } catch (error) {
      console.error('Search error:', error);
      res.status(500).json({ error: 'Internal server error' });
    }
  };
}